toolchain/check/import_ref.cpp

```diff
@@ -1432,10 +1260,42 @@ static auto TryResolveTypedInst(ImportRefResolver& resolver,
                          inst.adapted_type_inst_id, adapted_type_const_id));
 
   // Create a corresponding instruction to represent the declaration.
-  auto inst_id = AddImportedInst<SemIR::AdaptDecl>(
+  return ResolveAsUnique<SemIR::AdaptDecl>(
       resolver, import_inst_id, {.adapted_type_inst_id = adapted_type_inst_id});
-  return ResolveResult::Done(resolver.local_constant_values().Get(inst_id),
-                             inst_id);
+}
+
+static auto TryResolveTypedInst(ImportRefResolver& resolver,
+                                SemIR::AddrPattern inst,
+                                SemIR::InstId import_inst_id) -> ResolveResult {
+  auto type_const_id = GetLocalConstantId(resolver, inst.type_id);
+  auto inner_id = GetLocalConstantInstId(resolver, inst.inner_id);
+  if (resolver.HasNewWork()) {
+    return ResolveResult::Retry();
+  }
+
+  return ResolveAsUnique<SemIR::AddrPattern>(
+      resolver, import_inst_id,
+      {.type_id = resolver.local_context().types().GetTypeIdForTypeConstantId(
+           type_const_id),
+       .inner_id = inner_id});
+}
+
+template <typename ParamPatternT>
+  requires SemIR::HasInstCategory<SemIR::AnyParamPattern, ParamPatternT>
+static auto TryResolveTypedInst(ImportRefResolver& resolver, ParamPatternT inst,
```

Comment:
(No action requested, just musing)

It'd be nice to move away from using templating here and instead operate on `AnyParamPattern`s. We're pretty close to being able to do that -- I think what we're missing is a way to ask for the parse node kind and constant kind for an instruction category, so that `ResolveAsUnique` and `MakeImportedLocAndInst` can check it without knowing the exact instruction kind.